src/Database/LSMTree/Internal/BloomFilter.hs

```diff
@@ -0,0 +1,68 @@
+{-# LANGUAGE BangPatterns #-}
+module Database.LSMTree.Internal.BloomFilter (
+  bloomFilterToBuilder,
+  bloomFilterFromSBS,
+) where
+
+import           Control.Monad (when)
+import           Data.Bits (unsafeShiftL, unsafeShiftR, (.&.))
+import qualified Data.BloomFilter as BF
+import qualified Data.BloomFilter.BitVec64 as BV64
+import qualified Data.ByteString.Builder as B
+import           Data.ByteString.Short (ShortByteString (SBS))
+import qualified Data.Primitive as P
+import           Data.Primitive.ByteArray (ByteArray (ByteArray))
+import qualified Data.Vector.Primitive as PV
+import           Data.Word (Word32, Word64)
+import           Database.LSMTree.Internal.ByteString (byteArrayFromTo)
+
+-- serializing
+-----------------------------------------------------------
+
+bloomFilterVersion :: Word32
+bloomFilterVersion = 0
+
+bloomFilterEndianess :: Word32
+bloomFilterEndianess = 0  -- little endian
+
+bloomFilterToBuilder :: BF.Bloom a -> B.Builder
+bloomFilterToBuilder bf =
+    B.word32LE bloomFilterVersion <>
+    B.word32LE (fromIntegral (BF.hashesN bf)) <>
+    B.word32LE (fromIntegral (BF.length bf)) <>
+    B.word32LE bloomFilterEndianess <>
+    toBuilder' bf
+
+toBuilder' :: BF.Bloom a -> B.Builder
+toBuilder' (BF.B _hfN _len (BV64.BV64 (PV.Vector off len v))) = byteArrayFromTo (unsafeShiftL off 3) (unsafeShiftL off 3 + unsafeShiftL len 3) v
+
+-- deserializing
+-----------------------------------------------------------
+
+-- | Read 'BF.Bloom' from a 'ShortByteString'.
+--
+-- In successful case the data portion of bloom filter is /not/ copied (the short bytestring has only 16 bytes of extra data in the header).
+--
+bloomFilterFromSBS :: ShortByteString -> Either String (BF.Bloom a)
+bloomFilterFromSBS (SBS ba') = do
```

Comment:
Did you pick a non-incremental deserialisation function for a specific reason? An incremental version could, for example, use a `MutableByteArray` for the `vec64`, which we can copy chunks into from a lazy byte string

Reply:
You can read the whole bloom filter file into `MutableByteArray` directly (and freeze it). The file of the size is known.